k8s/sidecar/sidecar_webhook.go

```diff
@@ -0,0 +1,283 @@
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	logger.Info("Injecting vineyard sidecar container successfully!")
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledPod)
+}
+
+// BuildVineyardCommand builds the command for vineyardd image
+func (r *SidecarInjector) BuildVineyardCommand(sidecar *v1alpha1.Sidecar) string {
+	socket := "/var/run/vineyard.sock"
+	if sidecar.Spec.Volume.PvcName != "" {
+		socket = sidecar.Spec.Volume.MountPath + "/vineyard.sock"
+	}
+	command := "/usr/bin/wait-for-it.sh -t 60 etcd-for-vineyard." + sidecar.Namespace + ".svc.cluster.local:2379;" + "\n" +
+		"sleep 1;" + "\n" +
+		"/usr/local/bin/vineyardd" + "\n" +
+		"--sync_crds true" + "\n" +
+		"--socket " + socket + "\n" +
+		"--size " + sidecar.Spec.Size + "\n" +
+		"--stream_threshold " + strconv.Itoa(int(sidecar.Spec.StreamThreshold)) + "\n" +
+		"--etcd_cmd " + sidecar.Spec.EtcdCmd + "\n" +
+		"--etcd_prefix " + sidecar.Spec.EtcdPrefix + "\n" +
+		"--etcd_endpoint " + sidecar.Spec.EtcdEndpoint + "\n"
+
+	if sidecar.Spec.Spill.Path != "" {
+		command = command + "--spill_path " + sidecar.Spec.Spill.Path + "\n" +
+			"--spill_lower_rate " + sidecar.Spec.Spill.SpillLowerRate + "\n" +
+			"--spill_upper_rate " + sidecar.Spec.Spill.SpillUpperRate + "\n"
+	}
+
+	if sidecar.Spec.Metric.Enable {
+		command = command + "--metrics" + "\n" +
+			"-log_dir /var/log/vineyard/" + "\n"
+	}
+
+	return command
+}
+
+// ApplyCRToSidecar applies the sidecar cr to the pod
+func (r *SidecarInjector) ApplyCRToSidecar(sidecar *v1alpha1.Sidecar, pod *corev1.Pod) {
+	// add sleep to wait for the sidecar container to be ready
+	for i := range pod.Spec.Containers {
+		pod.Spec.Containers[i].Lifecycle = &corev1.Lifecycle{
+			PostStart: &corev1.LifecycleHandler{
+				Exec: &corev1.ExecAction{
+					Command: []string{
+						"/bin/sh",
+						"-c",
+						"sleep 120",
+					},
+				},
+			},
+		}
+	}
+	// add the sidecar container
+	sidecarContainer := corev1.Container{
+		Name:            "vineyard-sidecar",
+		Image:           sidecar.Spec.Image,
+		ImagePullPolicy: corev1.PullPolicy(sidecar.Spec.ImagePullPolicy),
+		Command:         []string{"/bin/bash", "-c", r.BuildVineyardCommand(sidecar)},
+		Env: []corev1.EnvVar{
+			{
+				Name: "VINEYARDD_UID",
+				ValueFrom: &corev1.EnvVarSource{
+					FieldRef: &corev1.ObjectFieldSelector{
+						FieldPath: "metadata.uid",
+					},
+				},
+			},
+			{
+				Name: "VINEYARDD_NAME",
+				ValueFrom: &corev1.EnvVarSource{
+					FieldRef: &corev1.ObjectFieldSelector{
+						FieldPath: "metadata.name",
+					},
+				},
+			},
+			{
+				Name: "VINEYARDD_NAMESPACE",
+				ValueFrom: &corev1.EnvVarSource{
+					FieldRef: &corev1.ObjectFieldSelector{
+						FieldPath: "metadata.namespace",
+					},
+				},
+			},
+		},
+		Ports: []corev1.ContainerPort{
+			{
+				Name:          "vineyard-rpc",
+				ContainerPort: 9600,
+				Protocol:      corev1.ProtocolTCP,
+			},
+		},
```

Comment:
Maybe you could have a yaml as template for vineyardd container, then you could be relieve from the error-prone hacks in `BuildVineyardCommand`.